src/core/Tree/Tree.tsx

```diff
@@ -0,0 +1,214 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+import React from 'react';
+import { CommonProps, useTheme } from '../utils';
+import '@itwin/itwinui-css/css/tree.css';
+import cx from 'classnames';
+
+export const TreeContext = React.createContext<
+  | {
+      nodeDepth: number;
+      subNodeIds?: string[];
+    }
+  | undefined
+>(undefined);
+
+export type NodeData<T> = {
+  subNodes?: Array<T>;
```

Comment:
To support lazy loading of sub-nodes, we need a flag that tells whether node has any subnodes like `hasSubnodes` or `isLeaf`.

Reply:
Added hasSubnodes